Baked depth cache improvements

daleeidd opened the pull request:
Continues on from issue #364

Overwriting is for quality of life. Otherwise, you would have to remove the texture property, bake, replace and finally delete the old texture. It can track path changes so users can rename GUID name if desired. GUIDs are used to avoid conflicts.

huwb commented:
Thanks for this!

I'm not 100% sure the best process here - tbh i'd happily merge this back right now since its going into a dev branch - but ill comment for now.

I mulled over this for a bit. I think we're on the right track but I've come to realise that "Populate on startup" is not logically equivalent to "bake cache on the fly". There are situations where a user may want a dynamic-rendered cache, but they don't want it populated on startup. They may populate it later on once all the sublevels have streamed in, etc.

In my code so far, if a saved cache is provided, it uses that, otherwise it relies on populate on startup, or it relies on the user populating the cache at some point. I think this may work, but it may not be the clearest thing in the world as the behaviour is implicit and it requires this knowledge/context.

One idea might be to have a drop down box to select the 'mode' of the depth cache, similar to how a reflection probe has different update modes. Then the user could either set it to e.g. "populate at runtime" or "baked offline". This would make things a bit more explicit?

daleeidd commented:
I think I understand. An enum called mode:
- "Populate at runtime" would be same behaviour as populate at startup?
- "Baked offline" would use the saved cache data?
- "Scripted" or "Manual" so developers can call it themselves?

huwb commented:
Yeah that was my line of thinking.

Just took a look at reflection probe and it seems to have relevant modes, might be good to align with this wherever it makes sense:

![image](https://user-images.githubusercontent.com/939901/71360108-98587200-2586-11ea-9d76-753438e2aba4.png)

daleeidd commented:
Oh right. I was looking at light probes. Oops. I will update the commit.

daleeidd commented:
Hopefully I have interpreted what you proposed correctly. I had to make some other changes so that it functioned as per selected options.

Also, feel free to go ahead and make any changes you want if it is more efficient than explaining what you want to me. I appreciate efficiency and I don't want to get in the way of that. I won't take it personally.

huwb commented:
Tried it out - seems to work well, great stuff!

I decided to try hiding the refresh mode if cache set to baked. It kind of escalated from there, it turns out almost every setting is for realtime and when set to baked it now collapses to this:

![image](https://user-images.githubusercontent.com/939901/71364969-892cf080-2595-11ea-9324-b647f57e61f7.png)

I dont love the complexity the new code adds but i think its probably ok. Let me know what you think? Feel free to experiment and change.

FYI EveryFrame is debug only because i dont know theres any point to update a cache each frame - if one wants a changing depth they can assign a "Register Ocean Depth Input" component to draw straight into the depth data and skip the middleman. So i dont think it should be added.

daleeidd commented:
Works great. Haha, yeah, I was trying to avoid going down that rabbit hole. Wish Unity would provide more attributes.

The way the quad was being initialised had it rendering a square depth into the ocean for on demand. Fixed.

I also added a button to populate the cache for on demand. Only a small bit of code but it is useful for testing etc.

huwb commented:
Super cool!

Made a couple of small additions myself.

So I think this is in good shape. The defaults should mean it continues working for existing users, so i'm fairly confident deploying this. Do you plan to make any other changes or finalisation?

daleeidd commented:
Neat. Nope, all good from me.

huwb commented:
All merged into master! Thanks this is a great contribution.
